File: Fyp/bidbazar/lib/data/models/product_model.dart
class productModel {
  String? sId;
  String? user;
  String? name;
  String? specs;

  int? price;
  bool? wishlist;
  List<String>? images;
  String? category;
  int? qty;
  int? soldqty;
  int? saleonprice;
  String? createdat;

  productModel(
      {this.sId,
      this.user,
      this.name,
      this.specs,
      this.price,
      this.wishlist,
      this.images,
      this.category,
      this.qty,
      this.saleonprice,
      this.soldqty,
        this.createdat
      });

  productModel.fromJson(Map<String, dynamic> json) {
    sId = json['_id'];
    user = json['user'];
    name = json['name'];
    specs = json['specs'];
    price = json['price'];
    wishlist = json['wishlist'];
    images = json['images'].cast<String>();
    category = json['category'];
    qty = json['qty'];
    soldqty = json['soldqty'];
    saleonprice = json['saleonprice'];
    createdat = json['createdat'] ?? "";

  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['_id'] = this.sId;
    data['user'] = this.user;
    data['name'] = this.name;
    data['specs'] = this.specs;
    data['price'] = this.price;
    data['wishlist'] = this.wishlist;
    data['images'] = this.images;
    data['category'] = this.category;
    data['qty'] = this.qty;
    data['soldqty'] = this.soldqty;
    data['saleonprice'] = this.saleonprice;
    return data;
  }
  
}

File: backend/src/models/product_model.js
const mongoose = require('mongoose');
const uuid = require('uuid');
const customDate = require('../controllers/date_controller');

const productSchema=mongoose.Schema({
    user:{
        type:mongoose.Schema.Types.ObjectId,
        ref:'User',
        required:true
    },
    name:{
        type:String,
        required:true
    },
    // model:{
    //     type:String,
    //     default: "not Entered"
    // },
    specs:{
        type:String,
        default: "not Entered"
    },
    price:{
        type:Number,
        required:true,
    },
    saleonprice:{//if price is heigher then sold the bid 
        type:Number,
        required:true,
    },
    wishlist:{
        type:Boolean,
        default:false,        
    },
    images:{
        type:Array,
        default:[]
    },
    category:{
        type:mongoose.Schema.ObjectId,
        ref:"Category", //model name as reference pass
        required:true,
    },
    qty:{
        type:Number,
        required:false,
        default:1
    },
    soldqty:{
        type:Number,
        required:false,
        default:0
    },
    createdat: {
        type:String,
        default: customDate.getStringFormattedDate()  // Set default value to current date and time
    },
    expire:{
        type:Boolean,
        default: false
    },
});


    // add pre-hook to mongoose schema methods to perform operation like save update

productSchema.pre("save",function(next){//save ke time kiya hoga  
    next();
});


productSchema.pre(['update','findOneAndUpdate','updateOne'],function(next){
    const update= this.getUpdate();
    delete update._id;
    next();
}
);



const productModel=mongoose.model('Product',productSchema);

module.exports=productModel;